rllib/core/rl_trainer/torch/torch_rl_trainer.py

```diff
@@ -1,10 +1,185 @@
+from typing import (
+    Any,
+    Mapping,
+    Union,
+    Type,
+    Sequence,
+    Hashable,
+    Optional,
+    Callable,
+)
+import torch
 
-logger = logging.getLogger(__name__)
+from ray.train.torch.train_loop_utils import _TorchAccelerator
+
+from ray.rllib.core.rl_module.rl_module import RLModule, ModuleID
+from ray.rllib.core.rl_trainer.rl_trainer import (
+    RLTrainer,
+    MultiAgentRLModule,
+    ParamOptimizerPairs,
+    Optimizer,
+    ParamType,
+    ParamDictType,
+)
+from ray.rllib.core.rl_module.torch.torch_rl_module import DDPRLModuleWrapper
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.typing import TensorType
+from ray.rllib.utils.nested_dict import NestedDict
 
-# TODO: Implement this
+logger = logging.getLogger(__name__)
 
 
 class TorchRLTrainer(RLTrainer):
-    pass
+    def __init__(
+        self,
+        module_class: Union[Type[RLModule], Type[MultiAgentRLModule]],
+        module_kwargs: Mapping[str, Any],
+        scaling_config: Mapping[str, Any],
+        optimizer_config: Mapping[str, Any],
+        distributed: bool = False,
+        in_test: bool = False,
+    ):
+        super().__init__(
+            module_class=module_class,
+            module_kwargs=module_kwargs,
+            scaling_config=scaling_config,
+            optimizer_config=optimizer_config,
+            distributed=distributed,
+            in_test=in_test,
+        )
+
+        self._world_size = scaling_config.get("num_workers", 1)
+        self._use_gpu = scaling_config.get("use_gpu", False)
+        # These attributes are set in the `build` method.
+        self._device = None
+
+    @property
+    @override(RLTrainer)
+    def module(self) -> MultiAgentRLModule:
+        return self._module
+
+    @override(RLTrainer)
+    def configure_optimizers(self) -> ParamOptimizerPairs:
+        lr = self.optimizer_config.get("lr", 1e-3)
+        return [
+            (
+                self.get_parameters(self._module[key]),
+                torch.optim.Adam(self.get_parameters(self._module[key]), lr=lr),
+            )
+            for key in self._module.keys()
+        ]
+
+    @override(RLTrainer)
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]]
+    ) -> ParamDictType:
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+        loss[self.TOTAL_LOSS_KEY].backward()
+        grads = {pid: p.grad for pid, p in self._params.items()}
+
+        return grads
+
+    @override(RLTrainer)
+    def apply_gradients(self, gradients: ParamDictType) -> None:
+
+        # make sure the parameters do not carry gradients on their own
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+
+        # set the gradient of the parameters
+        for pid, grad in gradients.items():
+            self._params[pid].grad = grad
```

Comment:
interesting. you guys decided to zero_grad(), do some processing on the grads then apply?
curious why we didn't modify the grads in-place?

Reply:
Hmm, probablby to cover the quirky case, where some parameters (that an optimizer is "responsible" for) are missing from the local variable `gradients` and thus would be updated by the optimizer, even though they shouldn't.
?

Reply:
@gjoliver It has to be generic for both torch and tf and any other auto-grad package. So the generic script is what you said, compute the gradients, process them if needed (e.g. grad clipping, etc), and then apply them via deep learning optimizers (e.g. Adam, Adagrad, etc). The reason we don't do it in-place is that only torch does that in-place and that's not a generic programming paradigm that can be assumed for other frameworks. If you also look at the API of `apply_gradients()` it take gradients as input, so it should use that variable and not assume anything about the parameters already having a .grad variable.

Reply:
I see. That makes sense.

Reply:
Ok, makes sense to me, too, then.